### PR TITLE
TensorFlow K.Function() additional ops via fetches

### PR DESCRIPTION
**Pull request by bzamecnik**
Allow passing fetches to tf.Session.run() in K.Function() in TensorFlow backend.

- allow passing additional operations, such as StagingArea.put()
- previously (#6693) only options and run_metadata could have been passed
- note that providing `feed_dict` substitutions doesn't make much sense here since:
  - they can be passed in the `input` argument
  - if they're passes via kwargs in model.compile() the values would be constant

The main reason is to allow using TensorFlow StagingArea for copying input data to GPU asynchronously. A working implementation of such as pipeline is in [StagingAreaCallback](https://github.com/rossumai/keras-multi-gpu/blob/master/keras_tf_multigpu/callbacks.py#L10). [Several examples are provided](https://github.com/rossumai/keras-multi-gpu/tree/master/keras_tf_multigpu/examples/bzamecnik/staging_area) (MNIST, CIFAR10, ResNet50.

Example usage (pseudocode):

```
features_tensor, labels_tensor = ... next batch ...
area = tf.contrib.staging.StagingArea(
    dtypes=[tf.float32, tf.float32])
area_put = area.put([features_tensor, labels_tensor])
area_get_features, area_get_labels = area.get()

input = Model(tensor=area_get_features)
model = ... create model ...
model.compile(optimizer='sgd', loss='categorical_crossentropy',
      target_tensors=[area_get_labels], fetches=[area_put])
```

This approach is inspired by https://github.com/avolkov1/keras_experiments/blob/master/keras_exp/_patch_tf_backend.py by @avolkov1, but the additional operations are passed uniformly via session_kwargs and there's better documentation.

### CODE REVIEW

**Comment by TimZaman** [comment]
Lol, duplicate of #8285 ! Merge whichever, they are practically feature compatible.

**Comment by bzamecnik** [comment]
@TimZaman Aah :) In the morning I've just made a PR from the branch that was waiting there for a week since at night I managed to make a working use case for that. Seems you made a PR earlier in the morning :)

The PRs are basically about the same thing. When looking at your commit, passing `feed_dict` via `session_kwargs` is nicely uniform, but it's not usable in `model.compile()`, since we'd provide constant values (8593309).

**Comment by TimZaman** [comment]
I’m in Mountain View; did the change last night in this TZ. The feed dict
dictionary passed in is not (deep)copied, so you can still change the feed
dict contents? That would make a use case. It’s up to the user anyway.

**Comment by TimZaman** [comment]
Lets merge yours, as yours has a test :-) i’ll close mine.

**Comment by bzamecnik** [comment]
OK. Aah, actually, that's a good point that `feed_dict` in this case is not deep copied. So that we can only pass a reference to the dict and then before each batch only change the value for the same placeholder key. I'll try to make a test for it and revert 8593309 if it works. Thanks!

**Comment by bzamecnik** [comment]
@TimZaman Thanks. It really works. See the unit test and an [example of usage](https://gist.github.com/bzamecnik/f76e480edf98e95ab263fd1a123af7a5).

**Comment by fchollet** [comment]
Ok, so should we merge this one?

File: keras/backend/tensorflow_backend.py [hunk 1]
```diff
@@ -2312,22 +2322,28 @@ def __init__(self, inputs, outputs, updates=None, name=None, **session_kwargs):
                     updates_ops.append(update)
             self.updates_op = tf.group(*updates_ops)
         self.name = name
+        # additional tensor substitutions
+        self.feed_dict = session_kwargs.pop('feed_dict', {})
+        # additional operations
+        self.fetches = session_kwargs.pop('fetches', [])
+        if not isinstance(self.fetches, list):
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Same comment as in the other PR: this could be any iterable, and we should normalize it to a list.

**Comment by ahundt** [comment]
This is also a duplicate of a PR I did 6 months ago, check how that compares as well https://github.com/fchollet/keras/pull/7064

File: keras/backend/tensorflow_backend.py (hunk 1 above)
Comment (bzamecnik) [reply keras/backend/tensorflow_backend.py]: In the `__call__()` method we construct a list of operations for `fetches` argument. `Session.run()` docs say it can contain an operation, list, tuple, dict or other structure of operations. Via `**session_kwargs` we should be able to pass anything that `run()` accepts. Also we ignore the results of these extra ops, since `__call__()` should return only results of `outputs` ops. Thus it seems to me that a correct behavior might be just to append `fetches` from `**session_kwargs` as an element of our constructed list and not try to convert it to list (ie. unwrap it). Trying to convert any iterable (including dict) to list would be wrong.

**Comment by ahundt** [comment]
Actually I think my PR will handle error situations with better notifications and error messages for the user, but @bzamecnik your unit test is better.

**Comment by ahundt** [comment]
I've merged https://github.com/fchollet/keras/pull/7064 with master and incorporated the unit test from this PR with credit to @bzamecnik. What do you think?

**Comment by ahundt** [comment]
Okay I compared them, if data isn't supplied correctly mine will raise this error while #8286  will not:
```
                raise ValueError('A tensor containing None '
                                 'was tied to value: ' + str(value) +
                                 ' so Session.run() cannot execute, '
                                 'please check your data and Model.')
```

However, mine is obviously more lines, and probably a touch slower. Up to you.

**Comment by TimZaman** [comment]
I think (this) Bohumir's MR is good to go.

**Comment by bzamecnik** [comment]
Thanks.

**Comment by ahundt** [comment]
cool, I'll close the other one.
